contracts/modules/Wallet/VestingEscrowWallet.sol

```diff
@@ -0,0 +1,525 @@
+pragma solidity ^0.4.24;
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/SafeERC20.sol";
+import "./IWallet.sol";
+
+/**
+ * @title Wallet for core vesting escrow functionality
+ */
+contract VestingEscrowWallet is IWallet {
+    using SafeMath for uint256;
+    using SafeERC20 for ERC20;
+
+    bytes32 public constant ADMIN = "ADMIN";
+
+    struct Schedule {
+        uint256 numberOfTokens;
+        uint256 lockedTokens;
+        uint256 duration;
+        uint256 frequency;
+        uint256 startTime;
+        uint256 nextTime;
+        State state;
+    }
+
+    struct Data {
+        uint256 index;
+        Schedule[] schedules;
+        uint256 availableTokens;
+        uint256 claimedTokens;
+    }
```

Comment:
I don't know why you make the data structure complex. it could be very simple something like that


        struct Schedule {
                 uint256 totalAmount;
                 uint256 claimedTokens;
                 uint256 duration;
                 uint256 frequency;
                 uint256 startTime;
           }
          no need of state as there is no operation depend on the state.
          mapping(address => Schedule[]); 

nothing else

Reply:
>no need of state as there is no operation depend on the state.
I've added state because we it in use case for Employee Vesting Dapp:
Employees want to see the status of the vesting schedules (i.e. to start, vesting, completed, terminated)

Reply:
I agree that state is redundant for now.

Reply:
I can change struct Schedule and get rid of struct Data, but in this case we will need to calculate availableTokens for some beneficiary on the fly. Is it ok?

Reply:
yes to calculate the tokens you can create some view function. and you can also create a view function to know the state of the schedule. no need to store explicitly and update everytime

              function getScheduleState(address _user, uint256 _index) external view returns(State) {
                         Schedule memory _userSchedule = schedules[_user][0];
                         if (now < _userSchedule.startTime)
                                return State.NOTSTARTED;
                         else if (now > _userSchedule.startTime && now < _userScheduleTime.startTime.add(duration) )
                                 return State.RUNNING
                          .
                          .
                          .
             }

Reply:
All changes have been made.
One difference: we need 3 variables with token numbers instead of 2.
    struct Schedule {
        uint256 numberOfTokens;
        uint256 releasedTokens;
        uint256 availableTokens;
        uint256 duration;
        uint256 frequency;
        uint256 startTime;
    }
numberOfTokens - number of tokens in creation time
releasedTokens - already released amount of tokens by schedule
availableTokens - available for withdrawing tokens

Reply:
> availableTokens - available for withdrawing tokens

can be calculated at execution time